tests/by-util/test_dmesg.rs

```diff
@@ -81,3 +81,111 @@ fn test_invalid_time_format() {
 }
+
+#[test]
+fn test_filter_facility() {
+    let facilities = [
+        "kern", "user", "mail", "daemon", "auth", "syslog", "lpr", "news", "uucp", "cron",
+        "authpriv", "ftp", "local0", "local1", "local2", "local3", "local4", "local5", "local6",
+        "local7",
+    ];
+    for facility in facilities {
+        let facility_filter_arg = format!("--facility={facility}");
+        let mut cmd = new_ucmd!();
+        let result = cmd
+            .arg("--kmsg-file")
+            .arg("kmsg.input")
+            .arg(facility_filter_arg)
+            .succeeds();
+        let stdout = result.no_stderr().stdout_str();
+        assert_eq!(stdout.lines().count(), 8);
+        let expected = format!("LOG_{}", facility.to_uppercase());
+        stdout
+            .lines()
+            .for_each(|line| assert!(line.contains(&expected)));
+    }
+}
+
+#[test]
+fn test_filter_levels() {
+    let levels = [
+        "emerg", "alert", "crit", "err", "warn", "notice", "info", "debug",
+    ];
+    for level in levels {
+        let level_filter_arg = format!("--level={level}");
+        let mut cmd = new_ucmd!();
+        let result = cmd
+            .arg("--kmsg-file")
+            .arg("kmsg.input")
+            .arg(level_filter_arg)
+            .succeeds();
+        let stdout = result.no_stderr().stdout_str();
+        assert_eq!(stdout.lines().count(), 20);
+        let expected = format!("LOG_{}", level.to_uppercase());
+        stdout
+            .lines()
+            .for_each(|line| assert!(line.contains(&expected)));
+    }
+}
+
+#[test]
+fn test_invalid_facility_argument() {
+    new_ucmd!()
+        .arg("--facility=definitely-invalid")
+        .fails()
+        .code_is(1)
+        .stderr_only("dmesg: unknown facility 'definitely-invalid'\n");
+}
+
+#[test]
+fn test_invalid_level_argument() {
+    new_ucmd!()
+        .arg("--level=definitely-invalid")
+        .fails()
+        .code_is(1)
+        .stderr_only("dmesg: unknown level 'definitely-invalid'\n");
+}
+
+#[test]
+fn test_filter_multiple() {
+    let mut cmd = new_ucmd!();
+    let result = cmd
+        .arg("--kmsg-file")
+        .arg("kmsg.input")
+        .arg("--facility=kern,user")
+        .arg("--level=emerg,alert")
+        .succeeds();
+    let stdout = result.no_stderr().stdout_str();
+    assert_eq!(stdout.lines().count(), 4);
+    stdout.lines().for_each(|line| {
+        assert!(
+            (line.contains("LOG_KERN") || line.contains("LOG_USER"))
+                && (line.contains("LOG_EMERG") || line.contains("LOG_ALERT"))
+        )
+    });
+}
+
+#[test]
+fn test_since_until() {
+    new_ucmd!()
+        .arg("--kmsg-file")
+        .arg("kmsg.input")
+        .arg("--since=\"2024-11-19 17:47:32 +0700\"")
+        .arg("--until='2024-11-19 18:55:52 +0700'")
+        .succeeds()
+        .no_stderr()
+        .stdout_is_templated_fixture("test_since_until.expected", &[("\r\n", "\n")]);
+}
```

Comment:
While the test passes, I noticed that when I run the command manually with `cargo run dmesg --kmsg-file tests/fixtures/dmesg/kmsg.input --since="2024-11-19 17:47:32 +0700" --until='2024-11-19 18:55:52 +0700'` there is no output.

Reply:
Yes, that is expected. To get the same output, you would need to modify `--since` to your boot time +80000 seconds and `--until` to your boot time +85000 seconds. You can get your boot time with `uptime -s`.

The reason for this is, `kmsg` file specify their time as relative to boot time. And for the test, I fixed the boot time (by enabling the `fixed-boot-time` feature) to my computer's boot time at the time of creating the `--time-format` test fixtures so that the tests are repeatable. You can see the fixed boot time in `time_formatter.rs:136`.

Maybe I should add an explanation comment on the test? e.g. "boot time +80000 s", "boot time + 85000 s", and "boot time fixed to 2024-11-18 19:34:12 with `--fixed-boot-time` feature enabled" or something like that.

Oh and actually, I miscalculated the `--until` argument, but it's not too far so the test still pass. I'll modify it too later.

One last note, there are multiple methods for getting the boot time, and all of them may not agree down to the seconds. To see what is the actual boot time used by our `dmesg`, you can call `cargo run dmesg -K tests/fixtures/dmesg/kmsg.input --time-format=iso | head`.

Reply:
Thanks for your explanation, that makes sense.